### ISSUE TITLE
Access part of a transient stream as a persistent one

### ISSUE THREAD

**Issue by frafra** [daggaz/json-stream#5]
I am trying to cycle over a list, but I need to access to some nested elements belonging to the same item multiple times. I do not want to store the whole JSON in memory, just the current element.

I tried to use `copy.copy` without success.
A sub-list can be stored using `.read_all()`, but not a dictionary inside a list.

```python
import json_stream
import io
from urllib.request import urlopen
import copy

url = 'https://api.ohsome.org/v1/elementsFullHistory/geometry?bboxes=9.188153743743898%2C45.463693737675456%2C9.193303585052492%2C45.464766009781414&properties=metadata&showMetadata=true&time=2007-10-08T00%3A00%3A00Z%2C2021-06-28T09%3A00%3A00Z&filter=type:node'

stream = json_stream.load(io.TextIOWrapper(urlopen(url), encoding='utf-8'))
for feature in stream['features']:
    f = copy.copy(feature)
    f['properties']['@osmId']
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/frafra/Code/is-osm-uptodate/venv/lib64/python3.9/site-packages/json_stream/base.py", line 35, in _iter_items
    item = self._load_item()
  File "/home/frafra/Code/is-osm-uptodate/venv/lib64/python3.9/site-packages/json_stream/base.py", line 170, in _load_item
    item = super()._load_item()
  File "/home/frafra/Code/is-osm-uptodate/venv/lib64/python3.9/site-packages/json_stream/base.py", line 130, in _load_item
    raise ValueError(f"Expecting value, comma or ], got {v}")
ValueError: Expecting value, comma or ], got }
```

**Comment by daggaz** [comment]
**Short answer**
You can't use `copy.copy()` (or `copy.deepcopy()`) on these objects. As stated in the docs for the `copy` module (emphasis mine):

> This module does not copy types like module, method, stack trace, stack frame, **file, socket**, window, array, or any similar types. It does “copy” functions and classes (shallow and deeply), by returning the original object unchanged; this is compatible with the way these are treated by the pickle module.

**Long answer**
When the iterator feeding the for loop yeilds an object it keeps a reference to that object (`self._child`). On the next iteration it then asks that child to consume any remaining data so the iteration can continue from the next list item (i.e. it calls `self._child.read_all()`).

When you copy the object, the state of the iterator and the stream get out of sync and cause the Exception.

**Very long answer**
More completely:
* the list iterator has just yeilded `feature` which it keeps a reference to (`self._child` representing the feature dict).
* you then `copy.copy()` that to a new object (also representing the feature dict)
* you operate on the copy and access `['properties']
* this creates another dict which the _copy_ keeps a reference to as its child (the original doesn't know about this)
* you access `['@osmId']` and get the data you want
* the loop body completes and the list iterator is asked to advance
* the list iterator asks its child (the original feature dict object) to clear itself
* the stream is not actually pointing within the feature dict, it's pointing within `feature['properties']` dict somewhere
* the original feature dict doesn't know this, because it doesn't have a child
* it therefore consumes the stream until it reaches the next `}` (the wrong one belonging to `feature['properties']`)
* the list then expects to be getting the next item (`,`) or the end of the list (`]`) but instead finds the closing `}` of the feature dict

**Note**
`read_all()` doesn't return anything, it just consumes the stream up to the end of the current item (list or dict).

For a list you can just call `list(streaming_list)`.

For a dict you can call `dict(streaming_dict.items())`

As you point out though, that will only give you one level, any nested data will be already consumed transient json objects which is not exactly helpful.

**Solution**
I think we need an additional API here to access part of a transient stream as a persistent one.

**Comment by daggaz** [comment]
Fixed in 1.1.0.

See the [docs](https://github.com/daggaz/json-stream#mixed-mode) on "mixed mode" for more info.
